src/haskell/duck.hs

```diff
@@ -0,0 +1,41 @@
+{- A typeclass that Wolf and Mallard will have to explicitly instantiate -}
+class Duck a where
+    quake :: a -> String
+    dance :: a -> String
+
+data Wolf = Wolf
+
+eat :: (Duck t) => Wolf -> t -> String
+eat _ _ = " 😈 "
+
+{- In order to become a duck, Wolf has to use the Adapter pattern. In
```

Comment:
I've seen about an equal number of people declaring that Haskell has or has no structural typing. The best answer I found is [here](http://stackoverflow.com/questions/21071603/why-does-haskell-not-have-records-with-structural-typing). It seems accepted that Haskell can provide a form of structural typing through the use of typeclasses : Mallard and DuckCosplay types can be considered as mere Duck because they instantiate it.
Apart from that, the compiler is very careful about types, which makes Haskell typing system sometimes cruelly nominal (e.g. won't compile when doing (1 :: Int) + (1 :: Integer) ).